oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/DefaultBearerTokenResolver.java

```diff
@@ -48,8 +48,9 @@
 
 	@Override
 	public String resolve(HttpServletRequest request) {
-		String authorizationHeaderToken = resolveFromAuthorizationHeader(request);
-		String parameterToken = resolveFromRequestParameters(request);
+		final String authorizationHeaderToken = resolveFromAuthorizationHeader(request);
+		final String parameterToken = isParameterTokenSupportedForRequest(request)
+				? resolveFromRequestParameters(request) : null;
```

Comment:
I've discussed this change with the team, and I have a suggestion for improving this a bit more (which I didn't think of when we first discussed this issue). For context, here is the [spec](https://datatracker.ietf.org/doc/html/rfc6750#section-2.2) verbiage:

> 2.2.  Form-Encoded Body Parameter
> 
>    When sending the access token in the HTTP request entity-body, the
>    client adds the access token to the request-body using the
>    "access_token" parameter.  The client MUST NOT use this method unless
>    all of the following conditions are met:
> 
>    o  The HTTP request entity-header includes the "Content-Type" header
>       field set to "application/x-www-form-urlencoded".
> 
>    o  The entity-body follows the encoding requirements of the
>       "application/x-www-form-urlencoded" content-type as defined by
>       HTML 4.01 [W3C.REC-html401-19991224].
> 
>    o  The HTTP request entity-body is single-part.
> 
>    o  The content to be encoded in the entity-body MUST consist entirely
>       of ASCII [USASCII] characters.
> 
>    o  The HTTP request method is one for which the request-body has
>       defined semantics.  In particular, this means that the "GET"
>       method MUST NOT be used.

It seems that all of those points are applicable to this scenario. In your case, you have a multi-part file upload (presumably with `Content-Type: multipart/form-data;`), which need not be concerned about accepting the bearer token as a request parameter in the body. To me, this means that we can split our token validation into two parts:
* `isParameterTokenSupportedForRequest(request)`
* `isParameterTokenEnabledForRequest(request)`

Something like:

```java
	private boolean isParameterTokenSupportedForRequest(HttpServletRequest request) {
		return (("POST".equals(request.getMethod())
				&& MediaType.APPLICATION_FORM_URLENCODED_VALUE.equals(request.getContentType()))
				|| "GET".equals(request.getMethod()));
	}

	private boolean isParameterTokenEnabledForRequest(HttpServletRequest request) {
		return ((this.allowFormEncodedBodyParameter && "POST".equals(request.getMethod())
				&& MediaType.APPLICATION_FORM_URLENCODED_VALUE.equals(request.getContentType()))
				|| (this.allowUriQueryParameter && "GET".equals(request.getMethod())));
	}
```

If you now check the new `isParameterTokenEnabledForRequest()` before returning the `parameterToken`, I think we should be able to avoid changing existing tests. We would want to add new tests for the `Content-Type: multipart/form-data;` and `Content-Type: application/x-www-form-urlencoded` cases specifically.

While we could also add a flag to enable/disable this new check (the new version of `isParameterTokenSupportedForRequest()`), I think we are fairly safe in the 5.6 release to state that we are enhancing spec compliance and security, and not make it configurable.

Reply:
Clarifying comment to above :wink: : "We should be able to avoid changing existing tests", with the exception of adding a `Content-Type` header to `POST` requests.